### DISCUSSION TITLE
How do I trigger a full refresh of incremental dbt models?

### DISCUSSION THREAD

**Discussion by james-larsen** [dagster-io/dagster#18880] [Q&A] [answered]
Hello all, we have a setup with Dagster orchestrating dbt in a Docker container.  I'm trying to define a job that will fully refresh all of our incremental jobs each morning.  From other help and resources I've received, it sounds like the best way is to pass a custom configuration into the job constructor.  Below is what I _**think**_ will build the job with our default configuration.  Can anyone offer any advice on how to tell Dagster to fully refresh all incremental models whenever that particular job is run?  Appreciate any help or insights!

``` python
dbt_asset_job_incremental_full_refresh = define_asset_job(
                               'dbt_job_incremental_full_refresh',
                                selection=AssetSelection
                                .keys(*incremental_asset_list)
                                .upstream(),
                                config=RunConfig(
                                    resources={
                                        "dbt": DbtCliResource(
                                            project_dir=DBT_PROJECT_PATH,
                                            profiles_dir=DBT_PROFILES,
                                        ),
                                        "slack": SlackResource(token=decoded_slack_token),
                                    }))
```

**Comment by rexledesma** [comment, marked as answer]
Hi James, this setup depends on how you've defined your dbt assets.

[In our API reference](https://docs.dagster.io/_apidocs/libraries/dagster-dbt#dagster_dbt.dbt_assets), we give an example of using `--full-refresh` with `@dbt_assets`. This uses [Dagster run configuration](https://docs.dagster.io/concepts/configuration/config-schema) to pass this custom configuration.

```python
from pathlib import Path

from dagster import AssetExecutionContext, Config
from dagster_dbt import DagsterDbtTranslator, DbtCliResource, dbt_assets


class MyDbtConfig(Config):
    full_refresh: bool


@dbt_assets(manifest=Path("target", "manifest.json"))
def my_dbt_assets(context: AssetExecutionContext, dbt: DbtCliResource, config: MyDbtConfig):
    dbt_build_args = ["build"]
    if config.full_refresh:
        dbt_build_args += ["--full-refresh"]

    yield from dbt.cli(dbt_build_args, context=context).stream()
```

After selecting these assets to execute in a job, the job will run `--full-refresh` if its config has set `full_refresh` to `True`.

```python
dbt_incremental_full_refresh_job = define_asset_job(
    "dbt_incremental_full_refresh_job",
    selection=AssetSelection(my_dbt_assets),
    config=RunConfig({"my_dbt_assets": MyDbtConfig(full_refresh=True)}),
)
```

**Reply by james-larsen** [reply]
Apologies for my ignorance and I really appreciate your help.  I'm new to Dagster / dbt and am not the one who set up the project, but I'm trying to get a POC working to present to the team.  Below is I think the part you are referring to.  Is there a way I can modify this so it doesn't impact the existing jobs that seem to use the default (none of the existing job constructors reference this asset or function)?  Or would it be better to create a second incremental_dbt_assets function to be used by that specific job?

```python
@dbt_assets(manifest=get_dbt_manifest(),
            dagster_dbt_translator=CustomDagsterDbtTranslator())
def dbt_assets(context: AssetExecutionContext, dbt: DbtCliResource):
    yield from dbt.cli(["run"], context=context).stream()
    yield from dbt.cli(["test"], context=context).stream()
```

**Reply by rexledesma** [reply]
You could just ensure that the default for `full_refresh` is `False`. Then existing invocations will function as before.

```python
class MyDbtConfig(Config):
    full_refresh: bool
    
@dbt_assets(manifest=get_dbt_manifest(), dagster_dbt_translator=CustomDagsterDbtTranslator())
def dbt_assets(context: AssetExecutionContext, dbt: DbtCliResource, config: MyDbtConfig):
    dbt_run_args = ["run"]
    if config.full_refresh:
        dbt_run_args += ["--full-refresh"]

    yield from dbt.cli(dbt_run_args, context=context).stream()
    yield from dbt.cli(["test"], context=context).stream()
```

**Reply by james-larsen** [reply]
This seems to have worked.  Thank you so much for your help!
